### PR TITLE
API definition and versioning

### PR DESCRIPTION
**Pull request by sebbalex**
This PR aims to:
- redefine the API specification
- use a valid API versioning model to guarantee backward compatibility
- refactor all code involved
- implement integration test for all API versions as well

### CODE REVIEW

File: apiv1/api.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+func parse(b []byte) ([]byte, error, error) {
+	url, err := utils.GetURLFromYMLBuffer(b)
+	if err != nil {
+		return nil, nil, err
+	}
+	p := publiccode.NewParser()
+	p.DisableNetwork = app.DisableNetwork
+
+	if url != nil {
+		p.DisableNetwork = app.DisableNetwork
```
Comment (bfabio) [inline apiv1/api.go:30]:
```suggestion
```

File: apiv1/api.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+	if err != nil {
+		app.DisableNetwork = false
+		log.Info("var disableNetwork not set, default to true")
+	}
+
+	Validate(w, r)
+}
+
+// Validate returns a YML or JSON onbject validated and upgraded
```
Comment (bfabio) [inline apiv1/api.go:156]:
```suggestion
// Validate returns a YML or JSON object validated and upgraded
```

**Review by bfabio — APPROVED** [review]
Not familiar with the code, but looks good AFAICT :sweat_smile:

File: utils/errors.go [hunk 3]
```diff
@@ -16,8 +16,8 @@ func (e ErrorInvalidKey) Error() string {
 
 // ErrorInvalidValue represents an error caused by an invalid value.
 type ErrorInvalidValue struct {
-	Key    string
-	Reason string
+	Key    string `json:"Key"`
```
Comment (ioggstream) [inline utils/errors.go:19]: @sebbalex Uppercase ok here?

Comment (sebbalex) [reply utils/errors.go:19]: Nope they are not, but we can't break the backward compatibility right now, they are going to be deprecated in next release.
Issue is still open to track this https://github.com/italia/publiccode-validator/issues/1#issuecomment-718094087

**Pull request merged**